scanners/csp.py

```diff
@@ -0,0 +1,71 @@
+import logging
+import requests
+from scanners import utils
+
+###
+# CSP Scanner - check the presence of CSP headers
+#
+
+
+# Set a default number of workers for a particular scan type.
+# Overridden by a --workers flag.
+workers = 2
+
+
+# default to a custom user agent, can be overridden
+user_agent = "github.com/18f/domain-scan, csp.py"
+
+
+def init_domain(domain, environment, options):
+    # If we have data from pshtt, skip if it's not a live domain.
+    if utils.domain_not_live(domain):
+        logging.debug("\tSkipping, domain not reachable during inspection.")
+        return False
+
+    # If we have data from pshtt, skip if it's just a redirector.
+    if utils.domain_is_redirect(domain):
+        logging.debug("\tSkipping, domain seen as just an external redirector during inspection.")
+        return False
+
+    # requests needs a URL, not just a domain.
+    url = None
+    if not (domain.startswith('http://') or domain.startswith('https://')):
+
+        # If we have data from pshtt, use the canonical endpoint.
+        if utils.domain_canonical(domain):
+            url = utils.domain_canonical(domain)
+
+        # Otherwise, well, ssl should work.
```

Comment:
You sure? :) It's usually safer to default to `http://` than `https://`, without information helping you pick which one.

Reply:
On this point I'm assuming that any uk gov.uk domain (my current daytime world) will be over TLS; the protection of data in transit is a core cloud security principle in the UK https://www.ncsc.gov.uk/guidance/cloud-security-principle-1-data-transit-protection but am happy to add a retry for first TLS and then plain text?

Reply:
I admire what the NCSC and GDS have been doing to encourage the use of encryption in the UK government, but I highly doubt that all gov.uk subdomains (and other UK government sites) will have HTTPS enabled. But I do suspect that most/all will redirect port 80 to 443. You'll likely get more complete data defaulting to port 80 and following redirects.

But this is also only for the case where no `pshtt` data exists. So you can ignore this issue entirely by including `pshtt` as a scan in your pipeline, for example running `./scan domain.gov.uk --scan=pshtt,third_parties`, which will allow `pshtt`'s calculation of the site's "canonical" URL to determine what URL gets scanned by `phantomas` here.